content/docs/middleware.mdx

```diff
@@ -0,0 +1,190 @@
+
+JSON Forms intercepts the call to the JSON Forms reducer and instead calls the provided middleware.
+The middleware acts as a reducer and can directly set the new state of JSON Froms. Warning: This deeply touches the JSONF Forms internals.
+The middleware, takes the current state and action as argument and returns the new state.
+In addition the JSON Forms default reducer is provided as agument:
+
+```js
+interface Middleware {
+  (
+    state: JsonFormsCore,
+    action: CoreActions,
+    defaultReducer: (state: JsonFormsCore, action: CoreActions) => JsonFormsCore
+  ): JsonFormsCore;
+}
+```
+
+The middleware allows to entirely change the behaviour of the reducer, make changes on top of the default reducer or provide custom implementations for selected actions, while using the default reducer in all other cases, like in the example below.
+
+```js
+const middleware = (
+    state: JsonFormsCore,
+    action: CoreActions,
+    defaultReducer: (state: JsonFormsCore, action: CoreActions) => JsonFormsCore
+  ) => {
+    switch (action.type) {
+        case INIT: {
+            //do whatever you like
+            const newState = customStateInit();
+            return newState
+        }
+        case UPDATE_DATA: {
+            const newState = defaultReducer(state, action);
+            update(newState);
+            return newState
+        }
+        default:
+            return defaultReducer(state, action);
+    }
+  }
+...
+
+<JsonForms
+  schema={schema}
+  uischema={uischema}
+  data={data}
+  renderers={renderers}
+  onChange={({ errors, data }) => setData(data)}
+  middleware
+/>
+```
+
+### Dependent Fields
+
+
+In the following example, we'll explore a scenario where one field depends on another.
+For instance, consider a carwash service that offers various services and calculates a price based on the selected options.
+We can utilize middleware to compute and establish the price. When an `UPDATE_DATA` action is triggered, we initially invoke the default middleware to update the data and identify any errors.
+Subsequently, we adjust the price fields based on the selected services and update the state with the newly calculated data.
+
+```js
+import { INIT, UPDATE_DATA } from  '@jsonforms/core'
+
+...
+  const [data, setData] = useState(inputData);
+
+  const middleware = useCallback((state, action, defaultMiddleware) => {
+    const newState = defaultMiddleware(state,action)
+    switch (action.type) {
+        case INIT:
+        case UPDATE_DATA: {
+            if(newState.data.services.length*15 !== newState.data.price){
+              newState.data.price =  newState.data.services.length*15;
+            }
+            setData(newState.data)
+            return newState;
+        }
+    default:
+        return newState
+    }
+  });
+
+...
+
+  return (
+    <JsonForms
+      data={data}
+      schema={schema}
+      renderers={materialRenderers}
+      middleware={middleware}
+    />
+  );
+```
+
+<DependentFieldExample/>
```

Comment:
Mention that if it's possible that new data is handed over to JSON Forms which does not follow this invariant, then the `UPDATE_CORE` action handling also needs to be overwritten.